src/samples/design_patterns/stream_sample.c

```diff
@@ -0,0 +1,285 @@
+            return false;
+        }
+    }
+
+    //Update how many bytes have been written
+    handle->written += status;
+    handle->pending_bytes -= status;
+
+    //Blob was completly written. Inform the user.
+    if (handle->written == blob->size) {
+        //Do we still have more data ?
+        sol_ptr_vector_del(&handle->pending_blobs, 0);
+        if (!sol_ptr_vector_get_len(&handle->pending_blobs)) {
+            r = false;
+            handle->write_monitor = NULL;
+        }
+
+        /*
+           Inform the user.
+           Since it's completly safe to call my_stream_api_close() inside tx_cb().
+           Informing the user should be the last thing to do.
+         */
+        if (handle->tx_cb)
+            handle->tx_cb((void *)handle->user_data, handle, blob, status);
+        sol_blob_unref(blob); //NOTE: that we unref the blob, not the user!
+    }
+
+    return r;
+}
+
+int
+my_stream_api_write(struct my_stream_api_handle *handle, struct sol_blob *blob)
+{
+    size_t total = handle->pending_bytes + blob->size;
+    int r;
+
+    //Tx_size was set and the total must not exceed tx_size
+    if (handle->tx_size && total >= handle->tx_size)
+        return -ENOSPC; //Try again later
+
+    //Store the blob and ref it, since it will written later.
+    r = sol_ptr_vector_append(&handle->pending_blobs, blob);
+    SOL_INT_CHECK(r, < 0, r);
+    sol_blob_ref(blob);
+    handle->pending_bytes = total;
+
+    //Schedule the write operation
+    if (!handle->write_monitor) {
+        handle->write_monitor = my_stream_device_add_io_monitor(handle->dev,
+            SOL_FD_FLAGS_OUT, _can_write, handle);
+        SOL_NULL_CHECK_GOTO(handle->write_monitor, err_monitor);
+    }
+
+    return 0;
+
+err_monitor:
+    sol_ptr_vector_del_element(&handle->pending_blobs, blob);
+    sol_blob_unref(blob);
+    return -ENOMEM;
+}
+
+//! [stream write]
+
+//! [stream read]
+//Delivery the data to the user
+static bool
+_inform_user(void *data)
+{
+    struct my_stream_api_handle *handle = data;
+    ssize_t r;
+
+    //Flag that we're using the handle, then if the user tries to delete us from the rx_cb we do not crash.
+    handle->in_use = true;
+
+    //Inform the user
+    r = handle->rx_cb((void *)handle->user_data, handle, &handle->rx);
+
+    handle->in_use = false;
+    //The user asked for deletion. Do it now.
+    if (handle->delete_me) {
+        handle->read_timeout = NULL;
+        my_stream_api_close(handle);
+        return false;
+    }
+
+    //Remove the data.
+    if (r < 0)
+        SOL_ERR("Something wrong happened %zd", r);
+    else
+        sol_buffer_remove_data(&handle->rx, 0, r);
+
+    if (handle->rx.used)
+        return true;
+    handle->read_timeout = NULL;
+    return false;
+}
+
+//Actually read from the device
+static bool
+_can_read(void *data, int fd, uint32_t active_flags)
+{
+    struct my_stream_api_handle *handle = data;
+    ssize_t status;
+
+    //Append data to the buffer
```

Comment:
I'd document the expand logic as you did in your real code, using fixed capacity checks